// dbs/dbs.js
let sqlite3 = require('sqlite3').verbose();
module.exports = {
  open(name) {
    let dbname = name || "':memory:'";
    return {
      handle: new sqlite3.Database(dbname),
      query(sql, params){
        let me = this;
        return new Promise((resolve, reject)=>{
          let recordset = {};
          recordset.error = null;
          recordset.items = [];
          me.handle.all(sql, params, (err, rows)=>{
            recordset.items = rows.concat();
            resolve(recordset);
          });
        });
      },
      run(sql){
        let me = this;
        return new Promise((resolve, reject)=>{
          me.handle.serialize(()=>{
            me.handle.run(sql); 
            resolve();
          });
        });
      },
      prepare(sql, args){
        let me = this;
        return new Promise((resolve, reject)=>{
          me.handle.serialize(()=>{
            let stmt = me.handle.prepare(sql);
            stmt.run(args);
            stmt.finalize();
            resolve();
          });
        }); 
      },
      close(){
        this.handle.close();
      }
    };
  }
};
